modules/@angular/compiler/src/metadata_resolver.ts

```
@@ -499,12 +499,16 @@ export class CompileMetadataResolver {
           transitiveModule.pipes.push(declaredIdentifier);
           declaredPipes.push(declaredIdentifier);
           this._addTypeToModule(declaredType, moduleType);
-        } else {
+        } else if (this._ngModuleResolver.isNgModule(declaredType)) {
```

Maybe don't add this extra case, but rather report an error `unexpected type ... Please add a @Pipe / @Directive / @Component annotation`.